src/Functions/FunctionsConversion.h

```diff
@@ -107,10 +107,20 @@ struct AccurateOrNullConvertStrategyAdditions
     UInt32 scale { 0 };
 };
 
+struct DateOrNullConvertStrategyAdditions
+{
+    UInt32 scale { 0 };
+};
 
 struct ConvertDefaultBehaviorTag {};
 struct ConvertReturnNullOnErrorTag {};
 
+template <typename FromType>
+static inline NO_SANITIZE_UNDEFINED bool isLtZero(const FromType & from)
```

Comment:
What about upper bound?
BTW, there are already methods for convertion, but their behaviour in case of overflow is to equal either lower or upper bound: https://github.com/ClickHouse/ClickHouse/blob/b024c215afa89fee914e44fded7000aa27c90995/src/Functions/FunctionsConversion.h#L352-L356
And a little lower - for date32 and dateTime and dateTIme64 from all integer types.
Probably you can just add template for overflow behaviour there (passing ConverExceptionMode, the default should be as it is now), which will be added only if there is suffix OrNull or OrZero?